sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/DeltaWriter.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.write;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * A data writer returned by {@link DeltaWriterFactory#createWriter(int, long)} and is
+ * responsible for writing a delta of rows.
+ *
+ * @since 3.2.0
+ */
+@Experimental
+public interface DeltaWriter<T> extends DataWriter<T> {
+  void delete(T id);
+  void update(T id, T row);
+  void insert(T row);
```

Comment:
I guess my one question here is why we have insert when were just delegating write to it by default? Maybe we could add some javadoc explaining the difference?

Reply:
I should have added more docs to methods. I'll fix it in the coming days. We don't really use `write` but it is available in the parent. I guess we either default like now or throw an unsupported exception. Maybe, an exception will be safer. 

Thoughts, @holdenk?